modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/table/Table.js

```diff
@@ -138,9 +138,10 @@ function Table({dataLoading, items, itemsActions, schema, style}) {
 							)}
 
 							{!!items.length &&
-								items.map((item) => {
+								items.map((item, i) => {
 									const itemId =
-										item[selectedItemsKey ?? 'id'];
+										item[selectedItemsKey ?? 'id'] ||
+										`item-${i}-${+new Date()}`;
```

Comment:
After thinking about this, I changed my mind. I don't think we should cover this case, throwing warning is a good behavior! Having an index as key is [an anti-pattern in React](https://reactjs.org/docs/lists-and-keys.html#keys). It is unstable, causing poor performance at best, and bugs at any modification of items. Conversely, setting timestamp as a key has the same problem, it is unstable.

To fix the error, I think we should one of these two things:
1. Add `id` to data. I don't think we need to add it as a column in view, just as a value to data, a `UserEntry` property added [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/provider/ClassicFDSDataProvider.java#L67-L69).
2. Set `emailAddress` as `selectedItemsKey`.

This is giving a stronger purpose to the `selectedItemsKey`. I kinda wish that we are calling this prop `primaryKey`, as the purpose goes beyond selection. It even has very similar purpose as primary key in databases, used for indexing.

Reply:
Agree about the anti-pattern but this resultant string should be stable, I mean unique. However, is not the solution we should aim to.

Solution 1 can work if well documented, assuming all entities will have an `id`, obviously, but I'll rather go for solution 2. In fact, was my first choice but, as you mention, use an attr called `selectedItemsKey` with no selection present in the UI is weird...

Another option I used in the past is a generator function used as a singleton for the whole component:

```
function* keyGenerator() {
  let id = 0
  while (true) {
    id += 1
    yield `idGeneratorId${id}`
  }
}
```

This way we do not worry about this and devs do not need to deal with React internals if they do not supply the right props in items.

What do you think?

Reply:
>Agree about the anti-pattern but this resultant string should be stable, I mean unique. 

It is unique, but it is not stable. By stable, I mean keeps its value between renders. Between renders React compares elements that changed and re-renders only those that did. When checking difference in `map`ped items, it relies on `key` for comparison. If `key` is a part of data, it will not change. But, index may change, and timestamp will certainly change. This will confuse React and it will now know what and how to update.

>Solution 1 can work if well documented, assuming all entities will have an `id`

I meant to add `id` to this specific `classic` table, a copy of `user.getId()`. 

It's an interesting idea to add `id` to all tables that don't have it, and have some generator for it :thinking: It would have to be in backend, and persisted in some way. It may be possible, but I think this goes beyond the scope of this task.

Reply:
> I mean keeps its value between renders
Oh, totally right, it should lead to a bad perf for sure, thxs for the tip!

> I meant to add id to this specific classic table, a copy of user.getId().
Yes, that's okay. I proceed with that and we'll see if we end up with this problem in the near future again

Thxs!